aea/helpers/dialogue/base.py

```diff
@@ -412,90 +395,145 @@ def last_message(self) -> Optional[Message]:
+        if self.is_empty:
+            return False
 
-        :return: True if empty, False otherwise
+        if not self.has_message_id(message.message_id):
+            return False
+
+        retrieved_message = self.get_message(message.message_id)
+        return message == retrieved_message
+
+    def has_message_id(self, message_id: int) -> bool:
         """
-        return len(self._outgoing_messages) == 0 and len(self._incoming_messages) == 0
+        Check whether a message with the supplied message id exists in this dialogue.
 
-    def update(self, message: Message) -> bool:
+        :param message_id: the message id
+        :return: True if message with that id exists in this dialogue, False otherwise
+        """
+        if self.is_empty:
+            return False
+
+        return 1 <= message_id <= self.last_message.message_id
+
+    def update(self, message: Message) -> None:
         """
         Extend the list of incoming/outgoing messages with 'message', if 'message' belongs to dialogue and is valid.
 
         :param message: a message to be added
         :return: True if message successfully added, false otherwise
         """
-        if not message.is_incoming:
+        if not message.has_sender:
             message.sender = self.agent_address
 
-        self.ensure_counterparty(message)
-
         if not self.is_belonging_to_dialogue(message):
-            return False
-
-        is_extendable = self.is_valid_next_message(message)
-        if is_extendable:
-            if message.is_incoming:
-                self._incoming_messages.extend([message])
-            else:
-                self._outgoing_messages.extend([message])
-        return is_extendable
+            raise InvalidDialogueMessage(
+                "The message {} does not belong to this dialogue."
+                "The dialogue reference of the message is {}, while the dialogue reference of the dialogue is {}".format(
+                    message.message_id,
+                    message.dialogue_reference,
+                    self.dialogue_label.dialogue_reference,
+                )
+            )
 
-    def ensure_counterparty(self, message: Message) -> None:
-        """
-        Ensure the counterparty is set (set if not) correctly.
+        is_valid_result, is_valid_message = self.is_valid_next_message(message)
 
-        :param message: a message
-        :return: None
-        """
-        counterparty = None  # type: Optional[str]
-        try:
-            counterparty = message.counterparty
-        except AssertionError:
-            # assume message belongs to dialogue
-            message.counterparty = self.dialogue_label.dialogue_opponent_addr
+        if not is_valid_result:
+            raise InvalidDialogueMessage(
+                "Message {} is invalid with respect to this dialogue. Error: ".format(
+                    message.message_id
+                )
+                + is_valid_message
```

Comment:
don't use `+`, you already use format, so add another `{}` and pass it to format

Reply:
Done